pkg/controller/workload/mpijob/mpijob_controller.go

```diff
@@ -102,6 +98,177 @@ func NewReconciler(
+		return nil
+	}
+	orderedReplicaTypes := orderedReplicaTypes(&job.Spec)
+	for index := range nodeSelectors {
+		replicaType := orderedReplicaTypes[index]
+		nodeSelector := nodeSelectors[index]
+		if len(nodeSelector) != 0 {
+			if job.Spec.MPIReplicaSpecs[replicaType].Template.Spec.NodeSelector == nil {
+				job.Spec.MPIReplicaSpecs[replicaType].Template.Spec.NodeSelector = nodeSelector
+			} else {
+				for k, v := range nodeSelector {
+					job.Spec.MPIReplicaSpecs[replicaType].Template.Spec.NodeSelector[k] = v
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (job *MPIJob) RestoreNodeAffinity(nodeSelectors []map[string]string) error {
+	orderedReplicaTypes := orderedReplicaTypes(&job.Spec)
+	for index := range nodeSelectors {
+		replicaType := orderedReplicaTypes[index]
+		nodeSelector := nodeSelectors[index]
+		if !equality.Semantic.DeepEqual(job.Spec.MPIReplicaSpecs[replicaType].Template.Spec.NodeSelector, nodeSelector) {
+			job.Spec.MPIReplicaSpecs[replicaType].Template.Spec.NodeSelector = map[string]string{}
+			for k, v := range nodeSelector {
+				job.Spec.MPIReplicaSpecs[replicaType].Template.Spec.NodeSelector[k] = v
+			}
+		}
+	}
+	return nil
+}
+
+func (job *MPIJob) Finished() (metav1.Condition, bool) {
+	var conditionType kubeflow.JobConditionType
+	var finished bool
+	for _, c := range job.Status.Conditions {
+		if (c.Type == kubeflow.JobSucceeded || c.Type == kubeflow.JobFailed) && c.Status == corev1.ConditionTrue {
+			conditionType = c.Type
+			finished = true
+			break
+		}
+	}
+
+	message := "Job finished successfully"
+	if conditionType == kubeflow.JobFailed {
+		message = "Job failed"
+	}
+	condition := metav1.Condition{
+		Type:    kueue.WorkloadFinished,
+		Status:  metav1.ConditionTrue,
+		Reason:  "JobFinished",
+		Message: message,
+	}
+	return condition, finished
+}
+
+func (job *MPIJob) EquivalentToWorkload(wl kueue.Workload) bool {
+	owner := metav1.GetControllerOf(&wl)
+	// Indexes don't work in unit tests, so we explicitly check for the
+	// owner here.
+	if owner.Name != job.Name {
+		return false
+	}
+
+	if len(wl.Spec.PodSets) != len(job.Spec.MPIReplicaSpecs) {
+		return false
+	}
+	for index, mpiReplicaType := range orderedReplicaTypes(&job.Spec) {
+		mpiReplicaSpec := job.Spec.MPIReplicaSpecs[mpiReplicaType]
+		if pointer.Int32Deref(mpiReplicaSpec.Replicas, 1) != wl.Spec.PodSets[index].Count {
+			return false
+		}
+		// nodeSelector may change, hence we are not checking for
+		// equality of the whole job.Spec.Template.Spec.
+		if !equality.Semantic.DeepEqual(mpiReplicaSpec.Template.Spec.InitContainers,
+			wl.Spec.PodSets[index].Template.Spec.InitContainers) {
+			return false
+		}
+		if !equality.Semantic.DeepEqual(mpiReplicaSpec.Template.Spec.Containers,
+			wl.Spec.PodSets[index].Template.Spec.Containers) {
+			return false
+		}
+	}
+	return true
+}
+
+func (job *MPIJob) PriorityClass() string {
+	return calcPriorityClassName(&job.MPIJob)
+}
+
+func (job *MPIJob) PodsReady() bool {
```

Comment:
Not quite related to this PR, if MPIJob JobRunning is False, and we disable the `WaitForPodsReady`, will we still admit the workload?

Reply:
I don't think I understand the question, can you maybe rephrase? Generally, `PodsReady` or `JobRunning` is `False` before admission, regardless of `WaitForPodsReady` being enabled.